Clarify existing DateTime docs and document limitations (#37)
neo4j-drivers/neo4j-drivers.github.io: packstream/packstream-specification-1.md

)
```

- The `tz_offset_seconds` specifies the offset in seconds from UTC.

> - The `seconds` are the above offset added to the seconds since the [Unix epoch](https://en.wikipedia.org/wiki/Epoch_(computing)).

Reviewer: I don't understand this description either :/

How about something along the lines of

> - The `seconds` and `nanoseconds` are the time elapsed since `1970-01-01 00:00:00` in the same timezone.

Alternatively

> - `seconds` and `nanoseconds` are computed by replacing (*not converting*) the timezone information of a given DateTime with `UTC` and computing its difference from (i.e., time elapsed since) [Unix epoch](https://en.wikipedia.org/wiki/Epoch_(computing)) (`1970-01-01T00:00:00Z`).

Author: I think we should mention Unix epoch/timestamp in there.

`1655472600` is the numbers of seconds since the Unix Epoch in `Fri Jun 17 2022 13:30:00 UTC`.
If the time zone offset is +1 hour, i.e. 3600 seconds, then the corresponding `DateTime` instance is as follows:
```
{
  seconds: 1655476200 # that's the Unix timestamp (1655472600) + the timezone offset (3600)
  nanoseconds: 0,
  tz_offset_seconds: 3600

> }

Reviewer: This sounds like it's the new format (Packstream V2). Or am I misreading this?

Author: The new format would be `1655472600` (actual Unix timestamp in seconds), not `1655476200=1655472600+3600` (Unix timestamp + offset).

)
```

- The `tz_offset_seconds` specifies the offset in seconds from UTC.
- The `seconds` elapsed since the [Unix epoch](https://en.wikipedia.org/wiki/Epoch_(computing)), often referred as a
  Unix timestamp, **added** to the above offset.

> - The `nanoseconds` are what remains after the last second of the datetime. The amount of nanoseconds ranges from 0 to

Reviewer: ```suggestion
- The `nanoseconds` are what remains after the last second of the DateTime. The amount of nanoseconds ranges from 0 to
```

> - The `nanoseconds` are what remains after the last second of the datetime. The amount of nanoseconds ranges from 0 to
>   999_999_999 (`_` separator added here and later for clarity).

Reviewer: ```suggestion
- The `nanoseconds` are what remains after the last second of the datetime. The amount of `nanoseconds` ranges from `0` to
  `999_999_999` (`_` separator added here and later for clarity).
```

For instance, the serialization of the point in time denoted as `1970-01-01T02:15:00+01:00` (and `42` nanoseconds) can 
be implemented as follows:

>  - compute the UTC time, i.e. `1970-01-01T01:15:00`.

Reviewer: ```suggestion
 - compute the UTC time, i.e. `1970-01-01T01:15:00Z` (`Z` denotes UTC)
```

 - compute the difference between that UTC time and the Unix epoch, which is 1h15, i.e. `8_100` seconds.

>  - add the offset of 1 hour, i.e. `3_600` seconds to the above difference, which yields `11_700` (`8_100`+`3_600`)

Reviewer: ```suggestion
 - add the offset of +1 hour, i.e. `3_600` seconds, to the above difference, which yields `11_700` (`8_100`+`3_600`)
```

The resulting `DateTime` instance is therefore as follows:
```
{
  seconds: 11700
  nanoseconds: 42,
  tz_offset_seconds: 3600
}
```

The deserialization of such a `DateTime` structure expectedly happens in reverse:

 - remove the offset from the `seconds` field, which gives here `8_100`

>  - instantiate the idiomatic equivalent of `DateTime` based on that Unix timestamp
>  - localize the resulting UTC `DateTime` to the timezone of the specified offset

Reviewer: ```suggestion
 - instantiate the idiomatic equivalent of `DateTime` based on that Unix timestamp giving `1970-01-01T01:15:00Z`
 - localize the resulting UTC `DateTime` to the timezone of the specified offset giving `1970-01-01T02:15:00+0100`
```

)
```

- The `tz_id` is an identifier for a specific time zone, such as `"Europe/Paris"`.
- The `seconds` are the offset corresponding to the above timezone at the given point in time, added to the seconds 

> since the [Unix epoch](https://en.wikipedia.org/wiki/Epoch_(computing)).

Reviewer: Similarly as above, I sadly don't understand this better than what we had before.

)
```

- The `tz_id` specifies the timezone name as understood
  by [the timezone database](https://en.wikipedia.org/wiki/Tz_database).
- The `seconds` elapsed since the [Unix epoch](https://en.wikipedia.org/wiki/Epoch_(computing)), often referred as a
  Unix timestamp, **added** to the offset derived from the named timezone and specified the point in time.

> - The `nanoseconds` are what remains after the last second of the datetime. The amount of nanoseconds ranges from 0 to
>   999_999_999 (`_` separator added here and later for clarity).

Reviewer: ```suggestion
- The `nanoseconds` are what remains after the last second of the datetime. The amount of nanoseconds ranges from `0` to
  `999_999_999` (`_` separator added here and later for clarity).
```

> For instance, the serialization of the point in time denoted as `1970-01-01T02:15:00 Europe/Paris` (and `42`

Reviewer: ```suggestion
For instance, the serialization of the point in time denoted as `1970-01-01T02:15:00+0100[Europe/Paris]` (and `42`
```

nanoseconds) can be implemented as follows:

- retrieve the offset of the named timezone for that point in time, here +1 hour, i.e. `3_600` seconds (the resolution
  is not always defined, read the following `Known Limitations` section to learn more)

> - compute the UTC time, i.e. `1970-01-01T01:15:00`.

Reviewer: ```suggestion
- compute the UTC time, i.e. `1970-01-01T01:15:00Z` (`Z` denotes UTC)
```

- compute the difference between that UTC time and the Unix epoch, which is 1h15, i.e. `8_100` seconds.

> - add the resolved offset of 1 hour, i.e. `3_600` seconds to the above difference, which yields `11_700` (`8_100`+`3_600`)

Reviewer: ```suggestion
- add the resolved offset of 1 hour, i.e. `3_600`, seconds to the above difference, which yields `11_700` (`8_100`+`3_600`)
```

The resulting `DateTime` instance is therefore as follows:
```
{
  seconds: 11700
  nanoseconds: 42,
  tz_id: "Europe/Paris"
}
```

The deserialization of such a `DateTime` structure expectedly happens in reverse:

> - retrieve the offset of the named timezone for that point in time, here +1 hour, i.e. `3_600` seconds (the resolution

Reviewer: This line is hand-wavy: we don't have a point in time yet, we just have seconds, nanoseconds, and a tz_id.

Author: I mean as a physical entity, the point in time that "exists" in the physical world

Reviewer: Sure, but you have not computed it yet. We are in the process of converting a number `seconds` into such a point that we then can use in what ever language this is being implemented in. In other words: how do I get that offset while just knowing a number of seconds and a timezene name. Feel free to answer in Go or any specific language for now.

Reviewer: To get to the core of the problem: this is exactly where the protocol is broken. Generally, you can't. So we should at least describe the "best effort" approach implemented in our drivers.

Author: I agree but I'm not sure how to describe with more precision since it is actually subject to ambiguities

  is not always defined, read the following `Known Limitations` to learn more)
- remove the resolved offset from the `seconds` field, which gives here `8_100`

> - instantiate the idiomatic equivalent of `DateTime` based on that Unix timestamp
> - localize the resulting UTC `DateTime` to the timezone of the specified offset

Reviewer: ```suggestion
- instantiate the idiomatic equivalent of `DateTime` based on that Unix timestamp giving `1970-01-01T01:15:00Z`
- localize the resulting UTC `DateTime` to the timezone of the specified offset giving `1970-01-01T02:15:00+0100[Europe/Paris]`
```

#### Known limitations

##### Accuracy

The resolution of offsets for a given time zone name and point in time is bound to the accuracy of the underlying
[timezone database](https://en.wikipedia.org/wiki/Tz_database). In
particular, [time zones before 1970](https://en.wikipedia.org/wiki/Tz_database#Data_before_1970) are not as well
specified.

Moreover, the offset resolution likely occurs both on the Bolt client side and Bolt server side. They each rely on a 

> different timezone database. That could lead to unwanted discrepancies if these copies are not in sync.

Reviewer 2: Perhaps make it more explicit: "if these two timezone databases are not in sync".

##### Time Shifts

Not all instances of `DateTimeZoneId` map to a single valid point in time.

1. During time shifts like going from 2AM to 3AM in a given day and timezone, 2:30AM e.g. does not happen.
2. Similarly, when going from 3AM to 2AM in a given day and timezone, 2:30AM happens twice.

In the first case, a `DateTimeZoneId` specifying a time between 2AM and 3AM does not correspond to any actual points in
time for that timezone and is invalid.

> In the second case, points in time between 2AM and 3AM happened twice, but with a different offset.

Reviewer: ```suggestion
In the second case, all points in the time between 2AM and 3AM exist twice, but with a different offset.
```
